flow360/component/simulation/operating_condition.py

```diff
@@ -1,60 +1,76 @@
-from typing import Optional, Union
+from typing import Optional, Tuple, Union
 
-import numpy as np
 import pydantic as pd
-from pydantic import validate_arguments
 
+import flow360.component.simulation.units as u
 from flow360.component.simulation.framework.base_model import Flow360BaseModel
+from flow360.component.simulation.framework.cache_model_base import CacheModelBase
+from flow360.component.simulation.models.material import Air, FluidMaterialTypes
+from flow360.component.simulation.unit_system import (
+    DensityType,
+    LengthType,
+    PressureType,
+    TemperatureType,
+    VelocityType,
+    ViscosityType,
+)
 
-VelocityVectorType = Unioin[VelocityType.Vector, Tuple[pd.StrictStr, pd.StrictStr, pd.StrictStr]]
+VelocityVectorType = Union[VelocityType.Vector, Tuple[pd.StrictStr, pd.StrictStr, pd.StrictStr]]
 
 
-class ThermalState(Flow360BaseModel):
-    temperature: TemperatureType.Positive = 288.15
-    density: DensityType.Positive = 1.225
-    material: FluidMaterialTypes = Air()
-    # TODO: special serializer
-    _altitude: Optional[LengthType.Positive] = None
-    _temperature_offset: Optional[TemperatureType.Positive] = None
+class ThermalStateCache(Flow360BaseModel):
+    altitude: Optional[LengthType.Positive] = None
+    temperature_offset: Optional[TemperatureType] = None
+
+
+class ThermalState(CacheModelBase):
+    temperature: TemperatureType.Positive = pd.Field(288.15 * u.K, frozen=True)
+    density: DensityType.Positive = pd.Field(1.225 * u.kg / u.m**3, frozen=True)
+    material: FluidMaterialTypes = pd.Field(Air(), frozen=True)
+    _cached: ThermalStateCache = ThermalStateCache()
 
-    @validate_arguments
     @classmethod
+    @pd.validate_call
     def from_standard_atmosphere(
-        cls, altitude: LengthType.Positive = 0, temperature_offset: TemperatureType = 0
+        cls, altitude: LengthType.Positive = 0 * u.m, temperature_offset: TemperatureType = 0 * u.K
     ):
         # TODO: add standard atmosphere implementation
-        density = 1.225
-        temperature = 288.15
+        density = 1.225 * u.kg / u.m**3
+        temperature = 288.15 * u.K
 
-        return cls(
+        state = cls(
             density=density,
             temperature=temperature,
             material=Air(),
         )
+        state._cached = ThermalStateCache(altitude=altitude, temperature_offset=temperature_offset)
+
+        return state
 
     @property
-    def altitude(self) -> LengthType.Positive:
-        return self._altitude
+    def altitude(self) -> Optional[LengthType.Positive]:
+        return self._cached.altitude
```

Comment:
Also would be nice if we tell the user that it was not given by user instead of return none?

Reply:
I think it's ok. If they haven't specified they shouldn't expect a value

Reply:
Yes but it would be still better to let them know why it is none.